[analyzer] Harden safeguards for Z3 query times

steakhal opened the pull request:
This patch is a functional change.
https://discourse.llvm.org/t/analyzer-rfc-taming-z3-query-times/79520

As a result of this patch, individual Z3 queries in refutation will be
bound by 300ms. Every report equivalence class will be processed in
at most 1 second.

The heuristic should have only really marginal observable impact -
except for the cases when we had big report eqclasses with long-running
(15s) Z3 queries, where previously CSA effectively halted.
After this patch, CSA will tackle such extreme cases as well.

steakhal committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.4

[skip ci]

steakhal committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

steakhal committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

steakhal committed:
Some minor cleanups

Created using spr 1.3.4

steakhal commented:
I've tested this change on 200+ projects, and the overall effect is:
 - 0 new issues appear
 - 276 issues disappear (because we drop the report eqclass)
 
 The three most affected checker categories (all of them are spread across usually 20+ projects):
  - null-deref-like diagnostics (74)
 - OOBv2 (70)
 - uninit-read-like (55)
 
 This underpins the non-intrusive nature of this change.

steakhal commented:
Ping. @NagyDonat @Xazax-hun @haoNoQ

steakhal commented on clang/include/clang/StaticAnalyzer/Core/BugReporter/Z3CrosscheckVisitor.h:
```diff
@@ -44,21 +47,43 @@ class Z3CrosscheckVisitor final : public BugReporterVisitor {
+
+private:
+  const AnalyzerOptions &Opts;
+  unsigned NumZ3QueriesDoneInEqClass = 0;
```
Hmm, I dont think I use this anymore.

NagyDonat commented on clang/lib/StaticAnalyzer/Core/Z3CrosscheckVisitor.cpp:
```diff
@@ -101,18 +123,37 @@ void Z3CrosscheckVisitor::Profile(llvm::FoldingSetNodeID &ID) const {
-    // For backward compatibility, let's accept the first timeout.
+  if (Query.IsSAT && Query.IsSAT.value()) {
+    ++NumTimesZ3QueryAcceptsReport;
+    return AcceptReport; // sat
```
```suggestion
    return AcceptReport;
```
This old comment doesn't add anything + the "unset" on the other branch was slightly incorrect.

NagyDonat commented on clang/unittests/StaticAnalyzer/Z3CrosscheckOracleTest.cpp:
```diff
@@ -17,43 +18,126 @@ using Z3Decision = Z3CrosscheckOracle::Z3Decision;
+  ASSERT_EQ(AcceptReport, interpretQueryResult({SAT, 200_ms, 1000_step}));
+}
+
+TEST_F(Z3CrosscheckOracleTest, RejectEQClassIfHitsRLimitTooManyTimes) {
```
The name `"HitsRLimitTooManyTimes"` is a leftover from the earlier plan where one `rlimit` overrun wasn't enough to discard the eqclass. (This also applies for the next TC.)

NagyDonat reviewed (approved):
Oops, I vaguely recalled that I already reviewed this, but it turns out I didn't. Anyway, the change LGTM, I only added a few minor remarks. (We already discussed a high-level overview of this on Discourse.)

steakhal committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

steakhal committed:
Implement move-only rule-of-5 for: - `Z3Config`

Created using spr 1.3.4

steakhal commented:
Ah, it mixed the messages... Here it is:

- Removed unused `NumZ3QueriesDoneInEqClass`
- Added a new `crosscheck-with-z3-eqclass-timeout-threshold` config option.
- Removed misleading `// unsat` comment on return statement.
- Removed unnecessary `// sat` comment on return statement.
- Adjusted the name of the two last unit-tests.

steakhal committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

steakhal committed:
Landed first commit

Created using spr 1.3.4

steakhal committed:
NFC

Created using spr 1.3.4

steakhal commented:
Reverting this patch as requested in this [comment](https://github.com/llvm/llvm-project/pull/95128#issuecomment-2176008007).
